### PR TITLE
Add method to compute higher-order derivatives

### PR DESCRIPTION
**Pull request by joehuchette**
Thoughts?

### CODE REVIEW

**Comment by blegat** [comment]
Are we sure that the type of p will not change in the loop leading to loss of performance ? I would probably use a recursive version instead like
```julia
diff(p, x, n) = diff(diff(p, x), x, n)
```

**Comment by joehuchette** [comment]
That's probably better

**Comment by rdeits** [comment]
@blegat I don't think that would help, since the compiler can't figure out how many recursive calls will happen (since that depends on the value of `n`). So if the return type of `diff(p, x, n)` depends on `n`, then the result will be type-unstable even in your implementation (caveat: I haven't actually tried this). 

The one case where your version would probably be better is if somehow `diff(p, x, n)` results in a different type than `p` but is always the same type for any `n`. I guess that could happen if multiplying the coefficient type by an integer resulted in a different coefficient type. That could actually be the case for JuMP variables. 

Ok, I think I've talked myself into agreeing with you.

**Comment by blegat** [comment]
@joehuchette I added some tests in 4452b97 and made some of your change also apply for `differentiate` with 2 arguments. The last remaining broken test is the `DomainError`, see the comments in `test/diff.jl`.

**Comment by blegat** [comment]
We are still missing a test with `diff(Monomial, ..., 0)` and `diff(MatPoly, ..., 0)` according to codecov

File: src/diff.jl [hunk 1]
```diff
@@ -37,3 +41,19 @@ differentiate(p::RationalPoly, x::PolyVar) = (differentiate(p.num, x) * p.den -
 function differentiate{C}(p::PolyType{C}, xs::Vector{PolyVar{C}})
     [differentiate(p, x) for x in xs]
 end
+
+# In Julia v0.5, it returns Any
+Base.promote_op{C}(differentiate, ::Type{PolyVar{C}}, ::Type{PolyVar{C}}) = Term{true, Int}
```
Comment (joehuchette) [inline src/diff.jl]: Should this be something like ``::typeof(differentiate)`` instead?

Comment (blegat) [inline src/diff.jl]: What do you mean ? I don't understand your notation :(

Comment (joehuchette) [inline src/diff.jl]: is it intended for ``Base.promote_op`` to dispatch on the function ``differentiate``? Right now it's just the name of an argument (might as well be ``foo``).

Comment (blegat) [inline src/diff.jl]: I don't know the details of `promote_op` but the first argument should be a function and it does inference to check what is the output type of this function with a given input type. Here `differentiate` is a variable that holds the function object.

Comment (joehuchette) [inline src/diff.jl]: My concern was that it's not used in the function body anywhere, which seemed strange if it wasn't being used for dispatch. I think it's fine, though; dispatch should work fine based on the other arguments.

Comment (blegat) [inline src/diff.jl]: Oh you are right, I see what you mean now!

Comment (blegat) [inline src/diff.jl]: This should be fixed in 528dae9

**Comment by joehuchette** [comment]
Merging cuz I want to use this :)

**Pull request merged**